diff --git a/makefile b/makefile
index 97e0f6b..3a9631d 100644
--- a/makefile
+++ b/makefile
@@ -29,8 +29,8 @@ BFILE=bin/$(NAME)
 EXEFILE=bin/$(NAME)_static
 
 #CONDITIONAL PATH DEFINITON
-system: DYN_LIBS=-lz -lpthread -lbz2 -llzma
-system: HTSSRC=/usr/local
+system: DYN_LIBS=-lz -lpthread -lbz2 -llzma -lcurl -lcrypto
+system: HTSSRC=/usr
 system: HTSLIB_INC=$(HTSSRC)/include/htslib
 system: HTSLIB_LIB=$(HTSSRC)/lib/libhts.a
 system: BOOST_INC=/usr/include
