This usecase describe how to run a demo GWAS analysis with [plink2](https://www.cog-genomics.org/plink/2.0/) and [regenie](https://rgcgithub.github.io/regenie/).
This will use genotype and phenotype data formatted according to [CoMorMent specification](../gwas/pheno_geno_specification.md),
and the helper [gwas.py](../gwas/gwas.py) script that reads the phenotype data,
extracts user-defined subset of phenotypes and covariates,
and prepares the scripts or SLURM jobs for ``plink2`` and ``regenie`` analysis.
In this demo we're using example data from [reference/examples/regenie](../reference/examples/regenie) folder.
Take a moment to look at the [phenotype file](../reference/examples/regenie/example_3chr.pheno) and it's [dictionary file](../reference/examples/regenie/example_3chr.pheno.dict) which will be used throughout this example.
For genetic data, we're using hard genotype calles in plink format, with ``n=500`` individuals ([example_3chr.fam](../reference/examples/regenie/example_3chr.fam)) and ``m=500`` SNPs across three chromosomes ([example_3chr.bim](../reference/examples/regenie/example_3chr.bim)). Btw, if you see ``Stored with Git LFS`` on some github pages, feel free to click ``View raw`` link - it should show the content of the file you're trying to see.

Now, to run this use case, just copy the [gwas.py](../gwas/gwas.py) script from ``$COMORMENT/containers/gwas/gwas.py`` into your current folder, and run the following commands (where ``run1`` gives example of case/control GWAS with plink2, while ``run2`` is an example for quantitative traits with regenie; these choices are independent - you could run case/control GWAS with regenie, and quantitative trait with plink2 by choosing --analysis argument accordingly; the meaning of the ``/REF`` and ``$SIF`` is explained in [Getting started](../README.md#getting-started) section of the main README file, as well as the way you are expected to setup the ``SINGULARITY_BIND`` variable):
```
singularity exec --home $PWD:/home $SIF/python3.sif python gwas.py gwas \
--argsfile /REF/examples/regenie/example_3chr.argsfile --covar PC1 PC2 BATCH \
--pheno CASE CASE2 --analysis plink2 --out run1

singularity exec --home $PWD:/home $SIF/python3.sif python gwas.py gwas \
--argsfile /REF/examples/regenie/example_3chr.argsfile --covar PC1 PC2 BATCH \
--pheno PHENO PHENO2 --analysis regenie --out run2
```
Off note, if you configured a local python3 environment (i.e. if you can use python without containers), and you have basic packages such as numpy, scipy and pandas, you may use ``gwas.py`` script directly - i.e. drop ``singularity exec --home $PWD:/home $SIF/python3.sif`` part of the above comand. Otherwise, we recommend to export ``$PYTHON`` variable as follows: ``export PYTHON="singularity exec --home $PWD:/home $SIF/python3.sif python"``, and then it e.g. like this: ``$PYTHON gwas.py ...``.

We're going to use ``--argsfile`` argument pointing to [example_3chr.argsfile](../reference/examples/regenie/example_3chr.argsfile) to specify some lengthy flags used across all invocations of the ``gwas.py`` scripts in this tutorial. It defines what phenotype file to use (``--pheno-file``), which chromosome labels to use (``--chr2use``), which genotype file to use in fitting the regenie model (``--bed-fit``) as well as genotype file to use when testing for associations (``--bed-test``); the ``--variance-standardize`` will apply linear transformation to all continuous phenotypes so that they became zero mean and unit variance, similar [--variance-standardize](https://www.cog-genomics.org/plink/2.0/data#variance_standardize) argument in plink2:
```
# example_3chr.argsfile
--pheno-file /REF/examples/regenie/example_3chr.pheno
--bed-fit /REF/examples/regenie/example_3chr
--bed-test /REF/examples/regenie/example_3chr
--chr2use 1-3
--variance-standardize
```

Take a look at the resulting [run1.log](gwas_demo/run1.log) and [run2.log](gwas_demo/run2.log), to see if gwas.py was executed as intended.
For this small-scale demo example, you could execute the actual GWAS locally on your machine as follows:

```
export REGENIE="singularity exec --home $PWD:/home $SIF/gwas.sif regenie"
export PLINK2="singularity exec --home $PWD:/home $SIF/gwas.sif plink2"
export PYTHON="singularity exec --home $PWD:/home $SIF/python3.sif python"

cat run1_cmd.sh | bash
cat run2_cmd.sh | bash
```

Otherwise you need to submit the SLURM jobs, generated by gwas.py script. There are two jobs for ``plink2`` analysis: ``run1_plink2.1.job`` and ``run1_plink2.2.job``, and three jobs for ``regenie`` analysis: ``run1_regenie.1.job``, ``run1_regenie.2.job``, ``run1_regenie.3.job`` (and similarly for ``run2``). These jobs must be executed in order, i.e. ``.2.job`` need to wait for ``.1.job``, and ``.3.job`` need to wait for ``.2.job``. You still can submit all jobs at once, but use SLURM's dependency management as described [here](https://stackoverflow.com/questions/19960332/use-slurm-job-id):
```
RES=$(sbatch run1_plink2.1.job)
RES=$(sbatch --dependency=afterany:${RES##* } run1_plink2.2.job)
RES=$(sbatch run2_regenie.1.job)
RES=$(sbatch --dependency=afterany:${RES##* } run2_regenie.2.job)
RES=$(sbatch --dependency=afterany:${RES##* } run2_regenie.3.job)
```

To customize parameters in the header of the slurm jobs, use ``--slurm-job-name``, ``--slurm-account``, ``--slurm-time``, ``--slurm-cpus-per-task``, ``--slurm-mem-per-cpu`` arguments of the ``gwas.py`` script (and let us know if there is anything else you need to customize!).
Further, you may need to customize ``--module-load`` argument, which by default loads ``singularity/3.7.1`` module.
Feel free to replace this with other version of singularity, or list multiple modules if you need to load something else in addition to singularity.
(a handy trick: if you want to explicily avoid loading the singularity module, because it's pre-installed, but don't need any other modules, you may add another irrelevant module just to overwrite the default ``--module-load`` argument).
Finally, you need to customize ``--comorment-folder`` folder containing a ``containers`` subfolder with a full copy of https://github.com/comorment/containers.

For more results, see [gwas_demo](gwas_demo) folder. Main results are the following GWAS summary statistics:
```
run1_CASE.plink2.gz
run1_CASE2.plink2.gz
run2_PHENO.regenie.gz
run2_PHENO2.regenie.gz
```
Each file is merged across all chromosomes, and has a minimal set of columns (``SNP, CHR, BP, A1, A2, N, Z, BETA, SE, PVAL``), as described in the [specification](../gwas/sumstats_specification.md).

It is also supported to run GWAS on dosages stored in BGEN format, instead of using hard call phenotypes from plink's bed/bim/fam format.
If you have genotypes formatted this way, the only change you need is to replace ``--bed-test`` with ``--bgen-test``,
as in this example: [example_3chr_bgen.argsfile](../reference/examples/regenie/example_3chr_bgen.argsfile).
It is expected that ``.bgen`` has corresponding ``.sample`` and ``.bgen.bgi`` files.

To see more info about ``gwas.py`` arguments, try this:
```
>singularity exec --home $PWD:/home $SIF/python3.sif python gwas.py gwas --help

usage: gwas.py gwas [-h] [--argsfile ARGSFILE] [--out OUT] [--log LOG] [--chr2use CHR2USE]
                    [--slurm-job-name SLURM_JOB_NAME] [--slurm-account SLURM_ACCOUNT]
                    [--slurm-time SLURM_TIME] [--slurm-cpus-per-task SLURM_CPUS_PER_TASK]
                    [--slurm-mem-per-cpu SLURM_MEM_PER_CPU]
                    [--module-load MODULE_LOAD [MODULE_LOAD ...]]
                    [--comorment-folder COMORMENT_FOLDER] [--singularity-bind SINGULARITY_BIND]
                    [--pheno-file PHENO_FILE] [--dict-file DICT_FILE] [--fam FAM]
                    [--bed-fit BED_FIT] [--bed-test BED_TEST] [--bgen-fit BGEN_FIT]
                    [--bgen-test BGEN_TEST] [--covar COVAR [COVAR ...]]
                    [--variance-standardize [VARIANCE_STANDARDIZE [VARIANCE_STANDARDIZE ...]]]
                    [--pheno PHENO [PHENO ...]] [--pheno-na-rep PHENO_NA_REP]
                    [--analysis {plink2,regenie} [{plink2,regenie} ...]]

(followed by description of all arguments - we don't copy this information here since gwas.py tool is being actively developed)
```

