This usecase describe how to run a demo GWAS analysis with ``plink2`` and ``regenie``.
This will use genotype and phenotype data formatted according to [CoMorMent specification](../gwas/pheno_geno_specification.md),
and the helper [gwas.py](../gwas/gwas.py) script that reads the phenotype data,
extracts user-defined subset of phenotypes and covariates,
and prepares the scripts or SLURM jobs for ``plink2`` and ``regenie`` analysis.
In this demo we're using example data from [reference/example/regenie] folder.
Take a moment to look at the [phenotype file](../reference/examples/regenie/example_3chr.pheno) and it's [dictionary file](../reference/examples/regenie/example_3chr.pheno.dict) which will be used throughout this example.
For genetic data, we're using hard genotype calles in plink format, with ``n=500`` individuals ([example_3chr.fam](../reference/examples/regenie/example_3chr.fam)) and ``m=500`` SNPs across three chromosomes ([example_3chr.bim](../reference/examples/regenie/example_3chr.bim)).

Copy ``gwas.py`` script into your current folder.
If you already configured python3 environment with basic packages such as numpy, scipy and pandas, you may use ``gwas.py`` script directly.
Otherwise, you may use it through ``python3.sif`` container. For example, you may show help about available arguments
```
>singularity exec --home $PWD:/home $SIF/python3.sif python gwas.py gwas --help

usage: gwas.py gwas [-h] [--argsfile ARGSFILE] [--out OUT] [--log LOG] [--chr2use CHR2USE]
                    [--pheno-file PHENO_FILE] [--dict-file DICT_FILE] [--fam FAM]
                    [--bed-fit BED_FIT] [--bed-test BED_TEST] [--covar COVAR [COVAR ...]]
                    [--pheno PHENO [PHENO ...]] [--pheno-na-rep PHENO_NA_REP]
                    [--analysis {plink2,regenie} [{plink2,regenie} ...]]
(followed by description of all arguments - we don't copy this information here since gwas.py tool is being actively developed)
```

We're going to use ``--argsfile`` argument pointint to [example_3chr.argsfile](../reference/examples/regenie/example_3chr.argsfile) to specify some lengthy flags used across all invocations of the ``gwas.py`` scripts in this tutorial. It defines what phenotype file to use (``--pheno-file``), which chromosome labels to use (``--chr2use``), which genotype file to use in fitting the regenie model (``--bed-fit``) as well as genotype file to use when testing for associations (``--bed-test``):
```
# example_3chr.argsfile
--pheno-file /REF/examples/regenie/example_3chr.pheno
--bed-fit /REF/examples/regenie/example_3chr
--bed-test /REF/examples/regenie/example_3chr
--chr2use 1-3
```

Now, we can prepare everything for genotype-phenotype association analysis as follows (``run1`` for case/control, ``run2`` for quantitative traits):
```
singularity exec --home $PWD:/home $SIF/python3.sif python gwas.py gwas --argsfile /REF/examples/regenie/example_3chr.argsfile --pheno CASE CASE2 --covar PC1 PC2 BATCH --out run1

singularity exec --home $PWD:/home $SIF/python3.sif python gwas.py gwas --argsfile /REF/examples/regenie/example_3chr.argsfile --pheno PHENO PHENO2 --covar PC1 PC2 BATCH --out run2
```

Take a look at the resulting [run1.log](gwas_demo/run1.log) and [run2.log](gwas_demo/run2.log), to see if gwas.py was executed as intended.
For this small-scale demo example, you could execute the actual GWAS locally on your machine as follows:

```
export REGENIE="singularity exec --home $PWD:/home $SIF/gwas.sif regenie"
export PLINK2="singularity exec --home $PWD:/home $SIF/gwas.sif plink2"
export PYTHON="singularity exec --home $PWD:/home $SIF/python3.sif python"

cat run1_cmd.sh | bash
cat run2_cmd.sh | bash
```

Otherwise you need to submit the SLURM jobs, generated by gwas.py script.

For more results, see [gwas_demo] folder.


