# PGS toolset

More on PGS: <https://choishingwan.github.io/PRS-Tutorial/>

## Codes

- ``README.md``: this file
- ``pgs/pgs.py``: Python class definitions setup to create bash commands for different PGS tools (``plink``, ``PRSice2``, ``LDpred2``, etc.)
- ``run_pgs_synthetic.py``: Python run script setup to run PGS tools on synthetic datas provided in this repository, mainly for testing purposes.
- ``run_pgs_w_QC.py``: Python run script setup to run PGS tools on datas provided in this repository, mainly for testing purposes. Performs basic QC steps.
- ``run_pgs_MoBa_*.py``: Python run scripts to run PGS tools on MoBa datas.
- ``pgs_exec.py``: ``gwas.py`` like command line tool that can be used to run, create bash and slurm job scripts for different PGS tools based on user input.
- ``vis_pgs_synthetic.ipynb``: Jupyter notebook plotting/comparing the output PGS scores generated by the ``run_pgs_synthetic.py`` script.
- ``vis_pgs_w_QC.ipynb``: Jupyter notebook plotting/comparing the output PGS scores generated by the ``run_pgs_w_QC.py`` script.
- ``start_jupyter_server.py``: start a Jupyter server using the ``python3.sif`` container (allows jupyter notebooks within VSCode with the Jupyter extension)
- ``config.yaml``: YAML file defining some parameters for Slurm jobscripts and PGS methods
- ``pgs_exec_example_*.sh``: Example bash scripts for ``pgs_exec.py``
- ``requirements.txt``: Python package requirements. Install with ``pip install -r requirements.txt``
- ``Rscripts/*.R``: misc. ``R`` scripts defining or being used by the PGS tools or optional QC steps.
- ``tests/``: directory for unit tests, executable with ``py.test -v tests`` in this directory

## Requirements

The basic requirements for running these codes (sans project specific genomics data) are the same as for the rest of this project, i.e., a basic Python environment and a working Singularity/Apptainer installation.

## Running the codes

Running these codes requires Python 3.6+ (tested/developed mainly using Python 3.9+) and a working Singularity/Apptainer installation.

### Python runtime scripts

#### ``run_pgs_synthetic.py``

Run PGS using PLINK, PRSice2 and LDpred2 on synthetic data:

```
$ python3 run_pgs_synthetic.py
environment variables in use:
        ROOT_DIR: /nrec/space/espenh
        CONTAINERS: /nrec/space/espenh/containers
...
# A tibble: 1 × 12
  r.squared adj.r.squ…¹ sigma stati…²   p.value    df logLik   AIC   BIC devia…³
      <dbl>       <dbl> <dbl>   <dbl>     <dbl> <dbl>  <dbl> <dbl> <dbl>   <dbl>
1     0.765       0.762 0.494    201. 3.47e-150     8  -354.  728.  770.    120.
# … with 2 more variables: df.residual <int>, nobs <int>, and abbreviated
#   variable names ¹​adj.r.squared, ²​statistic, ³​deviance
[1] "R2 (null model): 0.0250978564710413"
[1] "R2 (full model): 0.765387984851408"
```

> **_NOTE:_**  The script may require other packages than Python builtins, such as ``pandas`` and ``pyyaml``. Install these by issuing ``pip install <package>`` in the current Python environment. There is also ``pip install -r requirements.txt``

#### ``run_pgs_w_QC.py``

Run PGS using PLINK, PRSice2 and LDpred2 on tutorial data, which performs some basic QC steps on the data based on suggestions from this [tutorial](https://choishingwan.github.io/PRS-Tutorial/ldpred2/):

```
python3 run_pgs_w_QC.py
```

#### ``run_pgs_MoBa_*.py``

Run PGS using PLINK, PRSice2 and LDpred2, respectively on MoBa data (only on TSD)


### ``pgs_exec.py`` command line tool

Run, create bash and slurm job scripts for different PGS tools

Get a list of options:

```
$ python3 pgs_exec.py --help
usage: PGS [-h] [--method {plink,prsice2,ldpred2-inf,ldpred2-auto}] [--config CONFIG] [--sumstats-file SUMSTATS_FILE] [--pheno-file PHENO_FILE] [--phenotype PHENOTYPE]
           [--phenotype-class {CONTINUOUS,BINARY}] [--geno-file-prefix GENO_FILE_PREFIX] [--output-dir OUTPUT_DIR] [--runtype {sh,slurm,subprocess}]
           {plink,prsice2,ldpred2-inf,ldpred2-auto} ...

A pipeline for PGS analysis

positional arguments:
  {plink,prsice2,ldpred2-inf,ldpred2-auto}

optional arguments:
  -h, --help            show this help message and exit
  --method {plink,prsice2,ldpred2-inf,ldpred2-auto}
                        Method for PGS
  --config CONFIG       config YAML file
  --sumstats-file SUMSTATS_FILE
                        summary statistics file
  --pheno-file PHENO_FILE
                        phenotype file
  --phenotype PHENOTYPE
                        phenotype name (must be a column header in ``pheno_file``)
  --phenotype-class {CONTINUOUS,BINARY}
                        phenotype class
  --geno-file-prefix GENO_FILE_PREFIX
                        file path to .bed, .bim, .fam, etc. files
  --output-dir OUTPUT_DIR
                        Output file directory
  --runtype {sh,slurm,subprocess}
                        operation mode
```

Example w. PRSice2 as subprocess on synthetic dataset (``pgs_exec_example_1.sh``):

```
python3 pgs_exec.py \
    --sumstats-file /REF/examples/ldpred2/trait1.sumstats.gz \
    --pheno-file /REF/examples/ldpred2/simu.pheno \
    --phenotype trait1 \
    --geno-file-prefix /REF/examples/ldpred2/g1000_eur_chr21to22_hm3rnd1 \
    --output-dir output/PGS_synthetic_prsice2 \
    --runtype subprocess \
    'prsice2' \
    --covariate-file '/REF/examples/prsice2/EUR.cov' \
    --eigenvec-file 'output/PGS_synthetic_prsice2/g1000_eur_chr21to22_hm3rnd1.eigenvec'
```

> **_NOTE:_**  The last two lines will override settings for ``method: prsice2`` in ``config.yaml`` file, being parsed to the ``PRSice2.r`` script

Example w. LDpred2-auto via shell (``sh``) script on synthetic dataset (``pgs_exec_example_2.sh``):

```
python3 pgs_exec.py \
    --sumstats-file '/REF/examples/ldpred2/trait1.sumstats.gz' \
    --pheno-file '/REF/examples/ldpred2/simu.pheno' \
    --phenotype 'trait1' \
    --geno-file-prefix '/REF/examples/ldpred2/g1000_eur_chr21to22_hm3rnd1' \
    --output-dir 'output/PGS_synthetic_LDpred2_auto' \
    --runtype 'sh' \
    'ldpred2-auto' \
    'file-keep-snps' '/REF/hapmap3/w_hm3.justrs' \
    'fileGenoRDS' 'output/g1000_eur_chr21to22_hm3rnd1.rds' \
    'chr2use' '21,22' \
    'cores' 4
```

Which generates a shell script that can be run as

```
bash bash_scripts/ldpred-inf.sh
```

> **_NOTE_** Replacing ``--runtype 'sh'`` with ``--runtype 'slurm'`` and ``'ldpred2-inf'`` by ``'ldpred2-auto'`` generates a slurm jobscript using LDpred2-auto which can be submitted by issuing ``bash slurm_job_scripts/ldpred2-auto.job`` (cf. ``pgs_exec_example_3.sh``)
